### DISCUSSION TITLE
2.5.2. 第二节：如何变换 TensorIR 习题询问

### DISCUSSION THREAD

**Discussion by NotUr77** [mlc-ai/mlc-zh#189] [Q&A]
我使用的是google colab运行的ipynb文件
为什么我使用了parallel、vectorize 和 unroll后的运行时间反而更长了

这是我未使用parallel、vectorize 和 unroll的转换代码：
![image](https://github.com/mlc-ai/mlc-zh/assets/87253992/03f7ddba-3483-4f19-80b8-1c8afdfef64d)
这是转后的运行时间：
![image](https://github.com/mlc-ai/mlc-zh/assets/87253992/3e0a579b-81a7-4467-baee-94ccf78182e9)

这是我使用parallel、vectorize 和 unroll的转换代码：
![image](https://github.com/mlc-ai/mlc-zh/assets/87253992/eadd757d-7c7a-4cc0-bc81-c915edb73bf9)
这是转后的运行时间：
![image](https://github.com/mlc-ai/mlc-zh/assets/87253992/c368fd98-36ca-4604-8a77-32c67ceab472)

**Comment by LiZempher** [comment]
我复现了你的代码，
- 首先，你的时间差的太多了，我测了很多次，差距并不大。可能是因为电脑处理器问题
- 再就是，小批次实验误差是比较大的，而且这是比较不同算子优化方案之间的速度，有差异很正常。但是同原始方案比，都有所进步。

算子的parallel等操作在我的理解中，其实就是改变了数据的访问顺序，或是利用多线程等使得数据计算更快。
测试中注释掉了`reverse`和`decompose`和后面的`split`，这两个改动是最大的，影响也最大。我这样才达到你图二的一半速度
![image](https://github.com/mlc-ai/mlc-zh/assets/118066549/5c951d3f-01db-4e68-a05c-caffd326db30)
只进行reorder才达到你的图二速度15.5左右
我也是MLC小白，我不认为这是我们需要学习的重点
